Avoid doing event-related work for CSS transition & animation playback events when no event listener is registered

cdumez opened the pull request:
#### 43ea8744bc6065aad7ae5988e32d31d253905e5f
<pre>
Avoid doing event-related work for CSS transition &amp; animation playback events when no event listener is registered
<a href="https://bugs.webkit.org/show_bug.cgi?id=254454">https://bugs.webkit.org/show_bug.cgi?id=254454</a>
rdar://107077986

Reviewed by Antoine Quint.

Avoid doing event-related work for CSS transition &amp; animation playback events
when no event listener is registered for these.

We should probably do the same for CSS Animation events. I&apos;ll evaluate these
separately and follow-up.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/transition-events-with-document-change-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/transition-events-with-document-change.html: Added.
* Source/WebCore/animation/DeclarativeAnimation.cpp:
(WebCore::DeclarativeAnimation::invalidateDOMEvents):
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::removeReplacedAnimations):
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::cancel):
(WebCore::WebAnimation::finishNotificationSteps):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::addListenerTypeIfNeeded):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EventNames.h:
(WebCore::EventNames::isCSSTransitionEventType const):

Canonical link: <a href="https://commits.webkit.org/262123@main">https://commits.webkit.org/262123@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/30f4f624995a285945e4d7e5634b15b7acc126c0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/653 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/673 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/700 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/826 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/572 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/645 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/721 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/758 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/811 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/661 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/621 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/622 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/802 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/608 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/643 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/585 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/630 "8 failures") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/617 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/608 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/570 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/623 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/148 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/627 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

graouts reviewed (changes requested):
Thanks for doing this work, the A/B numbers you report suggest it's a great idea to do this. However, I have a few concerns:

1. this patch says it's about CSS Transitions but also deals with the `cancel` event which is an `AnimationPlaybackEvent`. I think this should either be strictly focused on `TransitionEvent` or include all `AnimationPlaybackEvent` types, including `finish` and `remove`. See https://drafts.csswg.org/web-animations-1/#animation-playback-events-section.
2. what about events for CSS Animations?
3. it doesn't seem obvious to me that we should track whether animation targets have registered event listeners on the `DocumentTimeline`. It would seem more logical to me to check this per `DeclarativeAnimation` in `DeclarativeAnimation::invalidateDOMEvents()`, and maybe have a flag on `Styleable` to track these. This will also apply for CSS Animation events. For `AnimationPlaybackEvent` the target is `WebAnimation`.
4. could we move the logic of `EventNames::isTransitionAnimationEventType()` to `CSSTransitionEvent`?

graouts commented on Source/WebCore/animation/DeclarativeAnimation.cpp:
```diff
@@ -318,33 +318,35 @@ void DeclarativeAnimation::invalidateDOMEvents(Seconds elapsedTime)
-            enqueueDOMEvent(eventNames().transitionendEvent, intervalStart, effectTimeAtEnd());
-        } else if ((!wasIdle && !wasAfter) && isIdle)
-            enqueueDOMEvent(eventNames().transitioncancelEvent, elapsedTime, elapsedTime);
+        if (m_owningElement->document().hasListenerType(Document::TRANSITION_ANIMATION_LISTENER)) {
```
Maybe add this condition to the `else if (is<CSSTransition>(this))` statement to avoid a single, big nested `if`?

graouts commented on Source/WebCore/dom/EventNames.h:
```diff
@@ -379,6 +379,7 @@ struct EventNames {
     bool isTouchScrollBlockingEventType(const AtomString& eventType) const;
     bool isMouseClickRelatedEventType(const AtomString& eventType) const;
     bool isMouseMoveRelatedEventType(const AtomString& eventType) const;
+    bool isTransitionAnimationEventType(const AtomString& eventType) const;
```
I suggest to use either `isCSSTransitionEventType` (matching our internal class name) or `isTransitionEventType` (matching the IDL interface name).

cdumez committed:
Avoid doing event-related work for CSS transition & animation playback events when no event listener is registered
https://bugs.webkit.org/show_bug.cgi?id=254454
rdar://107077986

Reviewed by Antoine Quint.

Avoid doing event-related work for CSS transition & animation playback events
when no event listener is registered for these.

We should probably do the same for CSS Animation events. I'll evaluate these
separately and follow-up.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/transition-events-with-document-change-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/transition-events-with-document-change.html: Added.
* Source/WebCore/animation/DeclarativeAnimation.cpp:
(WebCore::DeclarativeAnimation::invalidateDOMEvents):
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::removeReplacedAnimations):
* Source/WebCore/animation/WebAnimation.cpp:
(WebCore::WebAnimation::cancel):
(WebCore::WebAnimation::finishNotificationSteps):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::addListenerTypeIfNeeded):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EventNames.h:
(WebCore::EventNames::isCSSTransitionEventType const):

Canonical link: https://commits.webkit.org/262123@main
